algorithms/kernel/dtrees/gbt/classification/gbt_classification_train_dense_default_impl.i

```diff
@@ -63,54 +63,133 @@ public:
+                    gh[2 * sampleInd[i]] = sigm - y[sampleInd[i]]; //gradient
+                    gh[2 * sampleInd[i] + 1] = sigm * (algorithmFPType(1.0) - sigm); //hessian
+                }
+            }
+            else
             {
-                const algorithmFPType sigm = algorithmFPType(1.0) / (algorithmFPType(1.0) + exp[i]);
-                gh[2 * sampleInd[i]] = sigm - y[sampleInd[i]]; //gradient
-                gh[2 * sampleInd[i] + 1] = sigm * (algorithmFPType(1.0) - sigm); //hessian
+                PRAGMA_IVDEP
+                PRAGMA_VECTOR_ALWAYS
+                for(size_t i = 0; i < n; ++i)
+                {
+                    const auto sigm = algorithmFPType(1.0) / (algorithmFPType(1.0) + exp[i]);
+                    gh[2 * i] = sigm - y[i]; //gradient
+                    gh[2 * i + 1] = sigm * (algorithmFPType(1.0) - sigm); //hessian
+                }
             }
         }
         else
         {
-            PRAGMA_IVDEP
-            PRAGMA_VECTOR_ALWAYS
-            for(size_t i = 0; i < n; ++i)
+            const size_t nPerBlock = n / nThreads;
+            const size_t nSurplus = n % nThreads;
+            if(sampleInd)
+            {
+                daal::threader_for(nThreads, nThreads, [&](size_t iBlock)
+                {
+                    size_t start = iBlock + 1 > nSurplus ? nPerBlock * iBlock + nSurplus : (nPerBlock + 1) * iBlock;
+                    size_t end = iBlock + 1 > nSurplus ? start + nPerBlock : start + (nPerBlock + 1);
+                    PRAGMA_IVDEP
+                    PRAGMA_VECTOR_ALWAYS
+                    for (size_t i = start; i < end; i++)
+                    {
+                        exp[i] = -f[sampleInd[i]];
+                        /* make all values less than threshold as threshold value
+                        to fix slow work on vExp on large negative inputs */
+                        if(exp[i] < expThreshold)
+                            exp[i] = expThreshold;
+                    }
+                });
+            }
+            else
+            {
+                daal::threader_for(nThreads, nThreads, [&](size_t iBlock)
+                {
+                    size_t start = iBlock + 1 > nSurplus ? nPerBlock * iBlock + nSurplus : (nPerBlock + 1) * iBlock;
+                    size_t end = iBlock + 1 > nSurplus ? start + nPerBlock : start + (nPerBlock + 1);
+                    PRAGMA_IVDEP
+                    PRAGMA_VECTOR_ALWAYS
+                    for (size_t i = start; i < end; i++)
+                    {
+                        exp[i] = -f[i];
+                        /* make all values less than threshold as threshold value
+                        to fix slow work on vExp on large negative inputs */
+                        if(exp[i] < expThreshold)
+                            exp[i] = expThreshold;
+                    }
+                });
+            }
+            daal::internal::Math<algorithmFPType, cpu>::vExp(n, exp, exp);
```

Comment:
So, you go thought all elements in the array of size n in parallel_for.
And after this - you use exponent in sequential. Exp is quite more compute intensive than what you do in parallel regions. Why didn't you move this in parallel loop also?

You can do it by this way:
```cpp
daal::threader_for(nThreads, nThreads, [&](size_t iBlock)
{
    size_t start = iBlock + 1 > nSurplus ? nPerBlock * iBlock + nSurplus : (nPerBlock + 1) * iBlock;
    size_t end = iBlock + 1 > nSurplus ? start + nPerBlock : start + (nPerBlock + 1);
    PRAGMA_IVDEP
    PRAGMA_VECTOR_ALWAYS
    for (size_t i = start; i < end; i++)
    {
        exp[i] = -f[i];
        /* make all values less than threshold as threshold value
        to fix slow work on vExp on large negative inputs */
        if(exp[i] < expThreshold)
            exp[i] = expThreshold;
    }
    daal::internal::Math<algorithmFPType, cpu>::vExp(end-start, exp + start, exp + start);
});
```
Also, I recommend you in this case don't use blocking by nThreads, but by some fixed set of rows, for example - 512. Because in this case all data will be in L1 cache before Exp calling.